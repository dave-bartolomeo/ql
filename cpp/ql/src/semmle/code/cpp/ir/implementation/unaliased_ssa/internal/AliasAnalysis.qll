private import AliasAnalysisInternal
import cpp
private import InputIR
private import semmle.code.cpp.ir.internal.IntegerConstant as Ints

private import semmle.code.cpp.models.interfaces.Alias

private class IntValue = Ints::IntValue;

/**
 * Converts the bit count in `bits` to a byte count and a bit count in the form
 * bytes:bits.
 */
bindingset[bits]
string bitsToBytesAndBits(int bits) {
  result = (bits / 8).toString() + ":" + (bits % 8).toString()
}

/**
 * Gets a printable string for a bit offset with possibly unknown value.
 */
bindingset[bitOffset]
string getBitOffsetString(IntValue bitOffset) {
  if Ints::hasValue(bitOffset) then
    if bitOffset >= 0 then
      result = "+" + bitsToBytesAndBits(bitOffset)
    else
      result = "-" + bitsToBytesAndBits(Ints::neg(bitOffset))
  else
    result = "+?"
}

/**
 * Gets the offset of field `field` in bits.
 */
private IntValue getFieldBitOffset(Field field) {
  if (field instanceof BitField) then (
    result = Ints::add(Ints::mul(field.getByteOffset(), 8),
      field.(BitField).getBitOffset())
  )
  else (
    result = Ints::mul(field.getByteOffset(), 8)
  )
}

/**
 * Holds if the operand `tag` of instruction `instr` is used in a way that does
 * not result in any address held in that operand from escaping beyond the
 * instruction.
 */
private predicate operandIsConsumedWithoutEscaping(Operand operand) {
  // The source/destination address of a Load/Store does not escape (but the
  // loaded/stored value could).
  operand instanceof AddressOperand or
  exists (Instruction instr |
    instr = operand.getUseInstruction() and
    (
      // Neither operand of a Compare escapes.
      instr instanceof CompareInstruction or
      // Neither operand of a PointerDiff escapes.
      instr instanceof PointerDiffInstruction or
      // Converting an address to a `bool` does not escape the address.
      instr.(ConvertInstruction).getResultType() instanceof BoolType
    )
  ) or
  // Some standard function arguments never escape
  isNeverEscapesArgument(operand)
}

private predicate operandEscapesDomain(Operand operand) {
  not operandIsConsumedWithoutEscaping(operand) and
  not operandIsPropagated(operand, _) and
  not isArgumentForParameter(_, operand, _) and
  not isOnlyEscapesViaReturnArgument(operand) and
  not operand.getUseInstruction() instanceof ReturnValueInstruction and
  not operand instanceof PhiInputOperand
}

/**
 * If the result of instruction `instr` is an integer constant, returns the
 * value of that constant. Otherwise, returns unknown.
 */
IntValue getConstantValue(Instruction instr) {
  if instr instanceof IntegerConstantInstruction then
    result = instr.(IntegerConstantInstruction).getValue().toInt()
  else
    result = Ints::unknown()
}

/**
 * Computes the offset, in bits, by which the result of `instr` differs from the
 * pointer argument to `instr`, if that offset is a constant. Otherwise, returns
 * unknown.
 */
IntValue getPointerBitOffset(PointerOffsetInstruction instr) {
  exists(IntValue bitOffset |
    (
      bitOffset = Ints::mul(Ints::mul(getConstantValue(instr.getRight()),
        instr.getElementSize()), 8)
    ) and
    (
      instr instanceof PointerAddInstruction and result = bitOffset or
      instr instanceof PointerSubInstruction and result = Ints::neg(bitOffset)
    )
  )
}

/**
 * Holds if any address held in `operand` is propagated to the result of the instruction,
 * offset by the number of bits in `bitOffset`. If the address is propagated, but the offset is not
 * known to be a constant, then `bitOffset` is unknown.
 */
private predicate operandIsPropagated(Operand operand, IntValue bitOffset) {
  exists(Instruction instr |
    instr = operand.getUseInstruction() and
    not operand.isDefinitionInexact() and
    (
      // Converting to a non-virtual base class adds the offset of the base class.
      exists(ConvertToBaseInstruction convert |
        convert = instr and
        bitOffset = Ints::mul(convert.getDerivation().getByteOffset(), 8)
      ) or
      // Converting to a derived class subtracts the offset of the base class.
      exists(ConvertToDerivedInstruction convert |
        convert = instr and
        bitOffset = Ints::neg(Ints::mul(convert.getDerivation().getByteOffset(), 8))
      ) or
      // Converting to a virtual base class adds an unknown offset.
      (
        instr instanceof ConvertToVirtualBaseInstruction and
        bitOffset = Ints::unknown()
      ) or
      // Conversion to another pointer type propagates the source address.
      exists(ConvertInstruction convert, Type resultType |
        convert = instr and
        resultType = convert.getResultType() and
        (
          resultType instanceof PointerType or
          resultType instanceof Class  //REVIEW: Remove when all glvalues are pointers
        ) and
        bitOffset = 0
      ) or
      exists(PointerOffsetInstruction offsetInstr |
        // Adding an integer to or subtracting an integer from a pointer propagates
        // the address with an offset.
        operand = offsetInstr.getLeftOperand() and
        bitOffset = getPointerBitOffset(offsetInstr)
      ) or
      // Computing a field address from a pointer propagates the address plus the
      // offset of the field.
      bitOffset = getFieldBitOffset(instr.(FieldAddressInstruction).getField()) or
      // A copy propagates the source value.
      operand = instr.(CopyInstruction).getSourceValueOperand() and bitOffset = 0 or
      // Some functions are known to propagate an argument
      isAlwaysReturnedArgument(operand) and bitOffset = 0
    )
  )
}

/**
 * Equivalent to `operandIsPropagated`, except that it also considers the operand to be propagated by
 * a call to a function that always returns the argument given by the operand.
 */
private predicate operandIsPropagatedIncludingByCall(Operand operand, IntValue bitOffset) {
  operandIsPropagated(operand, bitOffset) or
  exists(CallInstruction ci, Instruction init |
    isArgumentForParameter(ci, operand, init) and
    resultReturned(init, bitOffset)
  )
}

private predicate operandEscapesNonReturn(Operand operand) {
  // The address is propagated to the result of the instruction, and that result itself is returned
  operandIsPropagated(operand, _) and resultEscapesNonReturn(operand.getUseInstruction())
  or
  // The operand is used in a function call which returns it, and the return value is then returned
  exists(CallInstruction ci, Instruction init |
    isArgumentForParameter(ci, operand, init) and
    (
      resultMayReachReturn(init) and
      resultEscapesNonReturn(ci)
      or
      resultEscapesNonReturn(init)
    )
  )
  or
  isOnlyEscapesViaReturnArgument(operand) and resultEscapesNonReturn(operand.getUseInstruction())
  or
  operand instanceof PhiInputOperand and
  resultEscapesNonReturn(operand.getUseInstruction())
  or
  operandEscapesDomain(operand)
}

private predicate operandMayReachReturn(Operand operand) {
  // The address is propagated to the result of the instruction, and that result itself is returned
  operandIsPropagated(operand, _) and
  resultMayReachReturn(operand.getUseInstruction())
  or
  // The operand is used in a function call which returns it, and the return value is then returned
  exists(CallInstruction ci, Instruction init |
    isArgumentForParameter(ci, operand, init) and
    resultMayReachReturn(init) and
    resultMayReachReturn(ci)
  )
  or
  // The address is returned
  operand.getUseInstruction() instanceof ReturnValueInstruction
  or
  isOnlyEscapesViaReturnArgument(operand) and resultMayReachReturn(operand.getUseInstruction())
  or
  operand instanceof PhiInputOperand and
  resultMayReachReturn(operand.getUseInstruction())
}

private predicate operandReturned(Operand operand, IntValue bitOffset) {
  // The address is propagated to the result of the instruction, and that result itself is returned
  exists(IntValue bitOffset1, IntValue bitOffset2 |
    operandIsPropagated(operand, bitOffset1) and
    resultReturned(operand.getUseInstruction(), bitOffset2) and
    bitOffset = Ints::add(bitOffset1, bitOffset2)
  )
  or
  // The operand is used in a function call which returns it, and the return value is then returned
  exists(CallInstruction ci, Instruction init, IntValue bitOffset1, IntValue bitOffset2 |
    isArgumentForParameter(ci, operand, init) and
    resultReturned(init, bitOffset1) and
    resultReturned(ci, bitOffset2) and
    bitOffset = Ints::add(bitOffset1, bitOffset2)
    
  )
  or
  // The address is returned
  operand.getUseInstruction() instanceof ReturnValueInstruction and
  bitOffset = 0
  or
  isOnlyEscapesViaReturnArgument(operand) and resultReturned(operand.getUseInstruction(), _) and
  bitOffset = Ints::unknown()
}

private predicate isArgumentForParameter(CallInstruction ci, Operand operand, Instruction init) {
  exists(Function f |
    ci = operand.getUseInstruction() and
    f = ci.getStaticCallTarget() and
    (
      init.(InitializeParameterInstruction).getParameter() = f.getParameter(operand.(PositionalArgumentOperand).getIndex())
      or
      init instanceof InitializeThisInstruction and
      init.getEnclosingFunction() = f and
      operand instanceof ThisArgumentOperand
    ) and
    not f.isVirtual() and
    not f instanceof AliasFunction
  )
}

private predicate isAlwaysReturnedArgument(Operand operand) {
  exists(AliasFunction f |
    f = operand.getUseInstruction().(CallInstruction).getStaticCallTarget() and
    f.parameterIsAlwaysReturned(operand.(PositionalArgumentOperand).getIndex())
  )
}

private predicate isOnlyEscapesViaReturnArgument(Operand operand) {
  exists(AliasFunction f |
    f = operand.getUseInstruction().(CallInstruction).getStaticCallTarget() and
    f.parameterEscapesOnlyViaReturn(operand.(PositionalArgumentOperand).getIndex())
  )
}

private predicate isNeverEscapesArgument(Operand operand) {
  exists(AliasFunction f |
    f = operand.getUseInstruction().(CallInstruction).getStaticCallTarget() and
    f.parameterNeverEscapes(operand.(PositionalArgumentOperand).getIndex())
  )
}

private predicate resultReturned(Instruction instr, IntValue bitOffset) {
  operandReturned(instr.getAUse(), bitOffset)
}

private predicate resultMayReachReturn(Instruction instr) {
  operandMayReachReturn(instr.getAUse())
}

/**
 * Holds if any address held in the result of instruction `instr` escapes
 * outside the domain of the analysis.
 */
private predicate resultEscapesNonReturn(Instruction instr) {
  // The result escapes if it has at least one use that escapes.
  operandEscapesNonReturn(instr.getAUse())
}

/**
 * Holds if the address of the specified local variable or parameter escapes the
 * domain of the analysis.
 */
private predicate automaticVariableAddressEscapes(IRAutomaticVariable var) {
  // The variable's address escapes if the result of any
  // VariableAddressInstruction that computes the variable's address escapes.
  exists(VariableAddressInstruction instr |
    instr.getVariable() = var and
    resultEscapesNonReturn(instr)
  )
}

/**
 * Holds if the address of the specified variable escapes the domain of the
 * analysis.
 */
predicate variableAddressEscapes(IRVariable var) {
  automaticVariableAddressEscapes(var.(IRAutomaticVariable)) or
  // All variables with static storage duration have their address escape.
  not var instanceof IRAutomaticVariable
}

/**
 * Holds if the value of the specified parameter escapes the domain of the analysis.
 */
predicate parameterValueEscapes(IRParameterVariable param) {
  // The parameter's value escapes if the result of the `InitializeParameter` instruction that initializes the variable
  // escapes.
  exists(InitializeParameterInstruction instr |
    instr.getVariable() = param and
    resultEscapesNonReturn(instr)
  )
}

/**
 * Holds if the address computed by `instruction` consists of the address computed by `base`, plus a bit
 * offset `bitOffset`. The offset need not be a constant. Does not hold if `base` itself consists of a base address
 * plus an offset. Thus, this predicate computes the most distant base address.
 */
private predicate resultPointsWithin(Instruction instruction, Instruction base, IntValue bitOffset) {
  exists(Operand operand, IntValue baseBitOffset, IntValue propagatedBitOffset |
    operand = instruction.getAnOperand() and
    // If an operand is propagated, then the result points to the same region,
    // offset by the bit offset from the propagation.
    operandIsPropagatedIncludingByCall(operand, propagatedBitOffset) and
    resultPointsWithin(operand.getDefinitionInstruction(), base, baseBitOffset) and
    bitOffset = Ints::add(baseBitOffset, propagatedBitOffset)
  ) or
  (
    not operandIsPropagatedIncludingByCall(instruction.getAnOperand(), _) and
    base = instruction and
    bitOffset = 0
  )
}

/**
 * Holds if the address used by `operand` consists of the address computed by `base`, plus a bit
 * offset `bitOffset`. The offset need not be a constant. Does not hold if `base` itself consists of a base address
 * plus an offset. Thus, this predicate computes the most distant base address.
 */
predicate operandPointsWithin(AddressOperand operand, Instruction base, IntValue bitOffset) {
  resultPointsWithin(operand.getDefinitionInstruction(), base, bitOffset)
}

/**
 * Holds if the address computed by `instruction` consists of the address computed by `base`, plus a
 * constant bit offset `bitOffset`. Does not hold if `base` itself consists of a base address plus a constant
 * offset. Thus, this predicate computes the most distant base address for which the offset is still a constant.
 */
private predicate resultBaseAddressAndConstantOffset(Instruction instruction, Instruction base, int bitOffset) {
  exists(Operand operand, IntValue baseBitOffset, IntValue propagatedBitOffset |
    operand = instruction.getAnOperand() and
    operandIsPropagated(operand, propagatedBitOffset) and
    Ints::hasValue(propagatedBitOffset) and
    resultBaseAddressAndConstantOffset(operand.getExactDefinitionInstruction(), base, baseBitOffset) and
    bitOffset = baseBitOffset + propagatedBitOffset
  ) or
  (
    not exists(Operand operand, IntValue propagatedBitOffset |
      operand = instruction.getAnOperand() and
      operandIsPropagated(operand, propagatedBitOffset) and
      Ints::hasValue(propagatedBitOffset)
    ) and
    base = instruction and
    bitOffset = 0
  )
}

/**
 * Holds if the address used by `operand` consists of the address computed by `base`, plus a
 * constant bit offset `bitOffset`. Does not hold if `base` itself consists of a base address plus a constant
 * offset. Thus, this predicate computes the most distant base address for which the offset is still a constant.
 */
predicate operandBaseAddressAndConstantOffset(AddressOperand operand, Instruction base, int bitOffset) {
  resultBaseAddressAndConstantOffset(operand.getDefinitionInstruction(), base, bitOffset)
}

predicate operandPointsToVariable(AddressOperand operand, IRVariable var, IntValue bitOffset) {
  exists(VariableAddressInstruction instr |
    operandPointsWithin(operand, instr, bitOffset) and
    var = instr.getVariable()
  )
}
