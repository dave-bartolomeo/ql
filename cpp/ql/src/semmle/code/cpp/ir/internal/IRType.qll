private import cpp
private import IntegerConstant
private import IRUtilities
private import semmle.code.cpp.ir.implementation.raw.internal.IRConstruction as IRConstruction

private int getPointerSize() {
  result = max(NullPointerType nullType | any() | nullType.getSize())
}

private class BlobType extends Type {
  BlobType() {
    exists(this.(Class).getSize()) or  // Only include complete classes
    this instanceof ArrayType or
    this instanceof PointerToMemberType or
    this instanceof GNUVectorType
  }
}

private predicate isRepresentableType(Type type) {
  exists(Type unspecifiedType |
    unspecifiedType = type.getUnspecifiedType() and
    (
      unspecifiedType instanceof VoidType or
      unspecifiedType instanceof IntegralOrEnumType or  // Includes `bool`
      unspecifiedType instanceof FloatingPointType or
      unspecifiedType instanceof PointerType or
      unspecifiedType instanceof ReferenceType or
      unspecifiedType instanceof FunctionPointerIshType or
      unspecifiedType instanceof ErroneousType or
      unspecifiedType instanceof BlobType or
      unspecifiedType instanceof RoutineType
    )
  )
}

private newtype TIRType =
  TIRPRValueType(Type type) {
    isRepresentableType(type) and
    not type.getUnspecifiedType() instanceof RoutineType  // No such thing as a function prvalue
  } or
  TIRFunctionGLValueType() or
  TIRGLValueAddressType(Type type) {
    isRepresentableType(type)
  } or
  TIRUnknownBlobType(int byteSize) {
    IRConstruction::needsUnknownBlobType(byteSize)
  } or
  TIRUnknownType()

class IRType extends TIRType {
  string toString() {
    result = "???"
  }

  int getByteSize() {
    none()
  }

  predicate hasType(Type type, boolean isGLValue) {
    none()
  }
}

private class IRWrappedType extends IRType {
  Type ctype;

  IRWrappedType() {
    this = TIRPRValueType(ctype) or
    this = TIRGLValueAddressType(ctype)
  }
}

private class IRPRValueType extends IRWrappedType, TIRPRValueType {
  override final string toString() {
    result = ctype.getUnspecifiedType().toString()
  }

  override final int getByteSize() {
    result = ctype.getSize()
  }

  override final predicate hasType(Type type, boolean isGLValue) {
    type = ctype and
    isGLValue = false
  }
}

class IRErrorType extends IRPRValueType {
  IRErrorType() {
    ctype.getUnspecifiedType() instanceof ErroneousType
  }
}

class IRVoidType extends IRPRValueType {
  IRVoidType() {
    ctype.getUnspecifiedType() instanceof VoidType
  }
}

class IRNumericType extends IRPRValueType {
  IRNumericType() {
    exists(Type unspecifiedType |
      unspecifiedType = ctype.getUnspecifiedType() and
      (
        unspecifiedType instanceof FloatingPointType or
        unspecifiedType instanceof IntegralOrEnumType and not unspecifiedType instanceof BoolType
      )
    )
  }
}

class IRIntegerType extends IRNumericType {
  IRIntegerType() {
    ctype.getUnspecifiedType() instanceof IntegralOrEnumType
  }
}

private predicate isSignedInteger(Type type) {
  exists(IntegralType integralType |
    integralType = type.getUnspecifiedType() and
    integralType.isSigned()
  ) or
  exists(Enum enumType |
    enumType = type.getUnspecifiedType() and
    (
      enumType.getExplicitUnderlyingType().getUnspecifiedType().(IntegralType).isSigned() or
      not exists(enumType.getExplicitUnderlyingType())  // Assume signed.
    )
  )
}

class IRSignedIntegerType extends IRIntegerType {
  IRSignedIntegerType() {
    isSignedInteger(ctype)
  }
}

class IRUnsignedIntegerType extends IRIntegerType {
  IRUnsignedIntegerType() {
    not isSignedInteger(ctype)
  }
}

class IRFloatingPointType extends IRNumericType {
  IRFloatingPointType() {
    ctype.getUnspecifiedType() instanceof FloatingPointType
  }
}

class IRBooleanType extends IRPRValueType {
  IRBooleanType() {
    ctype.getUnspecifiedType() instanceof BoolType
  }
}

class IRBlobType extends IRType {
  int byteSize;

  IRBlobType() {
    exists(Type type |
      this = TIRPRValueType(type) and
      byteSize = type.getUnspecifiedType().(BlobType).getSize()
    ) or
    this = TIRUnknownBlobType(byteSize)
  }
}

private class IRWrappedBlobType extends IRBlobType, IRPRValueType {
}

private class IRUnknownBlobType extends IRBlobType, TIRUnknownBlobType {
  override final string toString() {
    result = "unknown[" + byteSize.toString() + "]"
  }

  override final int getByteSize() {
    result = byteSize
  }
}

private predicate isPointerIshType(Type type) {
  exists(Type unspecifiedType |
    unspecifiedType = type.getUnspecifiedType() and
    (
      unspecifiedType instanceof PointerType or
      unspecifiedType instanceof ReferenceType
    )
  )
}

class IRAddressType extends IRWrappedType {
  IRAddressType() {
    this = TIRGLValueAddressType(ctype) or
    isPointerIshType(ctype)
  }
}

private class IRPRValueAddressType extends IRAddressType, IRPRValueType {
}

private class IRGLValueAddressType extends IRAddressType, TIRGLValueAddressType {
  override final string toString() {
    result = "glval<" + ctype.getUnspecifiedType().toString() + ">"
  }
  
  override final int getByteSize() {
    result = getPointerSize()
  }

  override final predicate hasType(Type type, boolean isGLValue) {
    type = ctype and
    isGLValue = true
  }
}

class IRFunctionAddressType extends IRType {
  IRFunctionAddressType() {
    exists(Type ctype |
      this = TIRPRValueType(ctype) and
      ctype.getUnspecifiedType() instanceof FunctionPointerIshType
    ) or
    this = TIRFunctionGLValueType()
  }
}

private class IRFunctionPointerishType extends IRFunctionAddressType, IRPRValueType {
}

private class IRFunctionGLValueType extends IRFunctionAddressType, TIRFunctionGLValueType {
  override final string toString() {
    result = "glval<unknown>"
  }

  override final int getByteSize() {
    result = getPointerSize()
  }
}

class IRUnknownType extends IRType, TIRUnknownType {
  override final string toString() {
    result = "unknown"
  }
}

IRUnknownType getUnknownType() {
  any()
}

IRVoidType getVoidType() {
  exists(VoidType voidType |
    result.hasType(voidType, false)
  )
}

IRType getIRTypeForExpr(Expr expr) {
  exists(boolean isGLValue |
    (
      if expr.isGLValueCategory() then 
        isGLValue = true
      else
        isGLValue = false
    ) and
    result.hasType(expr.getType(), isGLValue)
  )
}

IRType getIRTypeForPRValue(Type type) {
  result.hasType(type, false)
}

IRAddressType getIRTypeForGLValue(Type type) {
  result.hasType(type, true)
}

IRSignedIntegerType getIntType() {
  exists(IntType type |
    type.isImplicitlySigned() and
    result.hasType(type, false)
  )
}

IRBooleanType getBoolType() {
  exists(BoolType type |
    result.hasType(type, false)
  )
}

IRFunctionGLValueType getFunctionGLValueType() {
  any()
}

IRUnknownBlobType getUnknownBlobType(int byteSize) {
  result.getByteSize() = byteSize
}
