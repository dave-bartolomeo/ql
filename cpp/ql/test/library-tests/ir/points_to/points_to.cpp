struct Point {
  int x;
  int y;
};

struct Base1 {
  int b1;
};

struct Base2 {
  int b2;
};

struct DerivedSI : Base1 {
  int dsi;
};

struct DerivedMI : Base1, Base2 {
  int dmi;
};

struct DerivedVI : virtual Base1 {
  int dvi;
};

void PointsTo(int a, Point& b, Point* c, int* d, DerivedSI* e, DerivedMI* f, DerivedVI* g) {
  int i = a;
  i = *&a;
  i = *(&a + 0);
  i = b.x;
  i = b.y;
  i = c->x;
  i = c->y;
  i = *d;
  i = *(d + 0);
  i = d[5];
  i = 5[d];
  i = d[a];
  i = a[d];

  int* p = &b.x;
  i = *p;
  p = &b.y;
  i = *p;
  p = &c->x;
  i = *p;
  p = &c->y;
  i = *p;
  p = &d[5];
  i = *p;
  p = &d[a];
  i = *p;

  Point* q = &c[a];
  i = q->x;
  i = q->y;

  i = e->b1;
  i = e->dsi;
  i = f->b1;
  i = f->b2;
  i = f->dmi;
  i = g->b1;
  i = g->dvi;
}