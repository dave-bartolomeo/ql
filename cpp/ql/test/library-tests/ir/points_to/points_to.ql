import default
import semmle.code.cpp.ir.implementation.unaliased_ssa.internal.SSAConstruction
private import semmle.code.cpp.ir.internal.IntegerConstant as Ints

module Raw {
  import semmle.code.cpp.ir.implementation.raw.IR
  import semmle.code.cpp.ir.implementation.raw.gvn.ValueNumbering
  import semmle.code.cpp.ir.implementation.unaliased_ssa.internal.AliasAnalysis
  import semmle.code.cpp.ir.implementation.unaliased_ssa.internal.SimpleSSA

  string pointsToText(AddressOperand operand) {
    exists(IRVariable var, Ints::IntValue bitOffset |
      operandPointsToVariable(operand, var, bitOffset) and
      result = var.toString() + getBitOffsetString(bitOffset)
    ) or
    not operandPointsToVariable(operand, _, _) and result = "unknown" 
  }
}

module USSA {
  import semmle.code.cpp.ir.implementation.unaliased_ssa.IR
  import semmle.code.cpp.ir.implementation.unaliased_ssa.gvn.ValueNumbering
  import semmle.code.cpp.ir.implementation.aliased_ssa.internal.AliasAnalysis
  import semmle.code.cpp.ir.implementation.aliased_ssa.internal.AliasedSSA

  string regionText(AddressOperand operand) {
    exists(MemoryRegion region |
      region = getOperandMemoryRegion(operand) and
      result = region.getExampleInstruction().getOperationString()
    ) or
    not exists(getOperandMemoryRegion(operand)) and result = "unknown"
  }

  string baseOffsetText(AddressOperand operand) {
    exists(ValueNumber base, int bitOffset |
      operandBaseAndConstantOffset(operand, base, bitOffset) and
      result = base.getExampleInstruction().getOperationString() + getBitOffsetString(bitOffset)
    ) or
    not operandBaseAndConstantOffset(operand, _, _) and result = "??"
  }
}

from Raw::AddressOperand rawOperand, USSA::AddressOperand ussaOperand, string rawPointsTo,
  string ussaRegion, string ussaBaseOffset
where
  rawOperand.getUseInstruction() = getOldInstruction(ussaOperand.getUseInstruction()) and
  rawPointsTo = Raw::pointsToText(rawOperand) and
  ussaRegion = USSA::regionText(ussaOperand) and
  ussaBaseOffset = USSA::baseOffsetText(ussaOperand) and
  ussaRegion != "VariableAddress[i]" and
  ussaRegion != "VariableAddress[p]" and
  ussaRegion != "VariableAddress[q]"
select rawOperand.getUseInstruction().getLocation().toString(), rawOperand.getUseInstruction().getOperationString(),
  rawPointsTo, ussaRegion, ussaBaseOffset