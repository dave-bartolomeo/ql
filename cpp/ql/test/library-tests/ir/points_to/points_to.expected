| points_to.cpp:26:19:26:19 | InitializeParameter[a] | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:26:29:26:29 | InitializeParameter[b] | b+0:0 | VariableAddress[b] | VariableAddress[b]+0:0 |
| points_to.cpp:26:39:26:39 | InitializeParameter[c] | c+0:0 | VariableAddress[c] | VariableAddress[c]+0:0 |
| points_to.cpp:26:47:26:47 | InitializeParameter[d] | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:26:61:26:61 | InitializeParameter[e] | e+0:0 | VariableAddress[e] | VariableAddress[e]+0:0 |
| points_to.cpp:26:75:26:75 | InitializeParameter[f] | f+0:0 | VariableAddress[f] | VariableAddress[f]+0:0 |
| points_to.cpp:26:89:26:89 | InitializeParameter[g] | g+0:0 | VariableAddress[g] | VariableAddress[g]+0:0 |
| points_to.cpp:27:11:27:11 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:28:7:28:9 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:29:7:29:15 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:30:7:30:7 | Load | b+0:0 | VariableAddress[b] | VariableAddress[b]+0:0 |
| points_to.cpp:30:9:30:9 | Load | unknown | InitializeParameter[b] | InitializeParameter[b]+0:0 |
| points_to.cpp:31:7:31:7 | Load | b+0:0 | VariableAddress[b] | VariableAddress[b]+0:0 |
| points_to.cpp:31:9:31:9 | Load | unknown | InitializeParameter[b] | InitializeParameter[b]+4:0 |
| points_to.cpp:32:7:32:7 | Load | c+0:0 | VariableAddress[c] | VariableAddress[c]+0:0 |
| points_to.cpp:32:10:32:10 | Load | unknown | InitializeParameter[c] | InitializeParameter[c]+0:0 |
| points_to.cpp:33:7:33:7 | Load | c+0:0 | VariableAddress[c] | VariableAddress[c]+0:0 |
| points_to.cpp:33:10:33:10 | Load | unknown | InitializeParameter[c] | InitializeParameter[c]+4:0 |
| points_to.cpp:34:7:34:8 | Load | unknown | InitializeParameter[d] | InitializeParameter[d]+0:0 |
| points_to.cpp:34:8:34:8 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:35:7:35:14 | Load | unknown | InitializeParameter[d] | InitializeParameter[d]+0:0 |
| points_to.cpp:35:9:35:9 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:36:7:36:7 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:36:7:36:10 | Load | unknown | InitializeParameter[d] | InitializeParameter[d]+20:0 |
| points_to.cpp:37:7:37:10 | Load | unknown | InitializeParameter[d] | InitializeParameter[d]+20:0 |
| points_to.cpp:37:9:37:9 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:38:7:38:7 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:38:7:38:10 | Load | unknown | InitializeParameter[d] | PointerAdd[4]+0:0 |
| points_to.cpp:38:9:38:9 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:39:7:39:7 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:39:7:39:10 | Load | unknown | InitializeParameter[d] | PointerAdd[4]+0:0 |
| points_to.cpp:39:9:39:9 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:41:13:41:13 | Load | b+0:0 | VariableAddress[b] | VariableAddress[b]+0:0 |
| points_to.cpp:42:7:42:8 | Load | unknown | InitializeParameter[b] | InitializeParameter[b]+0:0 |
| points_to.cpp:43:8:43:8 | Load | b+0:0 | VariableAddress[b] | VariableAddress[b]+0:0 |
| points_to.cpp:44:7:44:8 | Load | unknown | InitializeParameter[b] | InitializeParameter[b]+4:0 |
| points_to.cpp:45:8:45:8 | Load | c+0:0 | VariableAddress[c] | VariableAddress[c]+0:0 |
| points_to.cpp:46:7:46:8 | Load | unknown | InitializeParameter[c] | InitializeParameter[c]+0:0 |
| points_to.cpp:47:8:47:8 | Load | c+0:0 | VariableAddress[c] | VariableAddress[c]+0:0 |
| points_to.cpp:48:7:48:8 | Load | unknown | InitializeParameter[c] | InitializeParameter[c]+4:0 |
| points_to.cpp:49:8:49:8 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:50:7:50:8 | Load | unknown | InitializeParameter[d] | InitializeParameter[d]+20:0 |
| points_to.cpp:51:8:51:8 | Load | d+0:0 | VariableAddress[d] | VariableAddress[d]+0:0 |
| points_to.cpp:51:10:51:10 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:52:7:52:8 | Load | unknown | InitializeParameter[d] | PointerAdd[4]+0:0 |
| points_to.cpp:54:15:54:15 | Load | c+0:0 | VariableAddress[c] | VariableAddress[c]+0:0 |
| points_to.cpp:54:17:54:17 | Load | a+0:0 | VariableAddress[a] | VariableAddress[a]+0:0 |
| points_to.cpp:55:10:55:10 | Load | unknown | InitializeParameter[c] | PointerAdd[8]+0:0 |
| points_to.cpp:56:10:56:10 | Load | unknown | InitializeParameter[c] | PointerAdd[8]+4:0 |
| points_to.cpp:58:7:58:7 | Load | e+0:0 | VariableAddress[e] | VariableAddress[e]+0:0 |
| points_to.cpp:58:10:58:11 | Load | unknown | InitializeParameter[e] | InitializeParameter[e]+0:0 |
| points_to.cpp:59:7:59:7 | Load | e+0:0 | VariableAddress[e] | VariableAddress[e]+0:0 |
| points_to.cpp:59:10:59:12 | Load | unknown | InitializeParameter[e] | InitializeParameter[e]+4:0 |
| points_to.cpp:60:7:60:7 | Load | f+0:0 | VariableAddress[f] | VariableAddress[f]+0:0 |
| points_to.cpp:60:10:60:11 | Load | unknown | InitializeParameter[f] | InitializeParameter[f]+0:0 |
| points_to.cpp:61:7:61:7 | Load | f+0:0 | VariableAddress[f] | VariableAddress[f]+0:0 |
| points_to.cpp:61:10:61:11 | Load | unknown | InitializeParameter[f] | InitializeParameter[f]+4:0 |
| points_to.cpp:62:7:62:7 | Load | f+0:0 | VariableAddress[f] | VariableAddress[f]+0:0 |
| points_to.cpp:62:10:62:12 | Load | unknown | InitializeParameter[f] | InitializeParameter[f]+8:0 |
| points_to.cpp:63:7:63:7 | Load | g+0:0 | VariableAddress[g] | VariableAddress[g]+0:0 |
| points_to.cpp:63:10:63:11 | Load | unknown | InitializeParameter[g] | ConvertToVirtualBase[DerivedVI : Base1]+0:0 |
| points_to.cpp:64:7:64:7 | Load | g+0:0 | VariableAddress[g] | VariableAddress[g]+0:0 |
| points_to.cpp:64:10:64:12 | Load | unknown | InitializeParameter[g] | InitializeParameter[g]+8:0 |
